### DISCUSSION TITLE
IPsec_SA was not established successfully, only-UDP support

### DISCUSSION THREAD

**Discussion by XYQ202** [strongswan/strongswan#1149] [Q&A]
Hi, Team
     
There is a new problem. When the device initiates a security association to the gateway, the following log is printed during the second stage (IPSEC_SA) of establishing an IPsec tunnel. Please help to analyze the reason for this?
![image](https://user-images.githubusercontent.com/97773232/179481463-26ed1ddb-4c26-4bf5-bc05-f3ea260efb5c.png)

Why does it appear that only UDP-encap is supported?

**Comment by tobiasbrunner** [comment]
Because you apparently are using [kernel-libipsec](https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html) (you probably shouldn't). Since that plugin announces `KERNEL_REQUIRE_UDP_ENCAPSULATION` forcing UDP-encapsulation is enabled, by default. Unless you are connecting to a peer that doesn't support NAT-T. So either don't use _kernel-libipsec_ but just _kernel-netlink_ (recommended), or make sure NAT-T is supported so UPD-encapsulation can be used.

**Reply by XYQ202** [reply]
I have canceled the use of the plugin kernel-libipsec and used kernel-netlink, but this error will occur, can you help me analyze it?

```
root@au_i2c_nand_v2:/usr/etc/swanctl# ipsec up OptusIPSEC
initiating IKE_SA OptusIPSEC[1] to 192.168.201.33
generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (336 bytes)
received packet: from 192.168.201.33[500] to 10.201.189.34[500] (265 bytes)
parsed IKE_SA_INIT response 0 [ SA KE No CERTREQ N(FRAG_SUP) ]
selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_1024
received cert request for "C=AU, O=Optus, CN=LTE-CA"
sending cert request for "C=AU, O=Optus, CN=LTE-CA"
authentication of 'C=AU, O=Optus, CN=049162B0C182@mavenir.com' (myself) with RSA signature successful
sending end entity cert "C=AU, O=Optus, CN=049162B0C182@mavenir.com"
establishing CHILD_SA OptusIPSEC{1}
generating IKE_AUTH request 1 [ IDi CERT N(INIT_CONTACT) CERTREQ IDr AUTH SA TSi TSr N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
splitting IKE message (1456 bytes) into 3 fragments
generating IKE_AUTH request 1 [ EF(1/3) ]
generating IKE_AUTH request 1 [ EF(2/3) ]
generating IKE_AUTH request 1 [ EF(3/3) ]
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (548 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (548 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (516 bytes)
received packet: from 192.168.201.33[500] to 10.201.189.34[500] (1124 bytes)
parsed IKE_AUTH response 1 [ EF(1/2) ]
received fragment #1 of 2, waiting for complete IKE message
received packet: from 192.168.201.33[500] to 10.201.189.34[500] (388 bytes)
parsed IKE_AUTH response 1 [ EF(2/2) ]
received fragment #2 of 2, reassembled fragmented IKE message (1440 bytes)
parsed IKE_AUTH response 1 [ IDr CERT AUTH SA TSi TSr ]
received end entity cert "C=AU, ST=NSW, L=Sydney, O=Optus, OU=Lab, CN=o22rrlabsecgw03 LTE Certificate, E=support@optus.com.au"
  using certificate "C=AU, ST=NSW, L=Sydney, O=Optus, OU=Lab, CN=o22rrlabsecgw03 LTE Certificate, E=support@optus.com.au"
  using trusted ca certificate "C=AU, O=Optus, CN=LTE-CA"
checking certificate status of "C=AU, ST=NSW, L=Sydney, O=Optus, OU=Lab, CN=o22rrlabsecgw03 LTE Certificate, E=support@optus.com.au"
certificate status is not available
  reached self-signed root ca with a path length of 0
authentication of '192.168.201.33' with RSA signature successful
IKE_SA OptusIPSEC[1] established between 10.201.189.34[C=AU, O=Optus, CN=049162B0C182@mavenir.com]...192.168.201.33[192.168.201.33]
scheduling reauthentication in 10546s
maximum IKE_SA lifetime 11626s
selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/NO_EXT_SEQ
**received netlink error: Function not implemented (38)
unable to add SAD entry with SPI c5c184a9 (FAILED)
received netlink error: Function not implemented (38)
unable to add SAD entry with SPI 5f6ecc45 (FAILED)
unable to install inbound and outbound IPsec SA (SAD) in kernel**
failed to establish CHILD_SA, keeping IKE_SA
sending DELETE for ESP CHILD_SA with SPI c5c184a9
generating INFORMATIONAL request 2 [ D ]
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (80 bytes)
received packet: from 192.168.201.33[500] to 10.201.189.34[500] (80 bytes)
parsed INFORMATIONAL response 2 [ D ]
establishing connection 'OptusIPSEC' failed
```

**Reply by tobiasbrunner** [reply]
You might be missing [required kernel modules](https://docs.strongswan.org/docs/5.9/install/kernelModules.html) (also check if crypto modules are loaded).

**Reply by XYQ202** [reply]
> You might be missing [required kernel modules](https://docs.strongswan.org/docs/5.9/install/kernelModules.html) (also check if crypto modules are loaded).

The default file used in my strongswan.conf file. The plugin should be loaded normally. It seems that there is something wrong with your strongswan. I changed the digest algorithm SHA256 in the second stage to SHA1, and the tunnel can be successfully established without prompting netlink. mistake：
![1](https://user-images.githubusercontent.com/97773232/179720004-3ceb0b9e-b4b9-441f-a7f2-16bd76032494.png)
![2](https://user-images.githubusercontent.com/97773232/179720011-a2cf829d-e856-43b4-ac34-7e8c9abe6471.png)

The second stage in png1 uses SHA256, while the second stage in png2 uses SHA1.  
can you provide some help

**Reply by tobiasbrunner** [reply]
> The default file used in my strongswan.conf file.

That has nothing to do with the Linux kernel modules.

> The plugin should be loaded normally.

Again, plugins in strongSwan are not kernel modules.

> I changed the digest algorithm SHA256 in the second stage to SHA1, and the tunnel can be successfully established without prompting netlink.

Sounds like your kernel is lacking support for SHA256 (or maybe configurable truncation, but I think only very old kernels are affected by that).
